clang/test/SemaSYCL/intel-max-global-work-dim-device.cpp

```diff
@@ -47,14 +61,109 @@ struct TRIFuncObjGood2 {
   operator()() const {}
 };
 
+struct TRIFuncObjGood3 {
+  [[intel::reqd_work_group_size(1, 1, 1)]]
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjGood4 {
+  [[cl::reqd_work_group_size(1, 1, 1)]]
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjGood5 {
+  [[intel::max_work_group_size(1, 1, 1)]]
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjGood6 {
+  [[intel::reqd_work_group_size(4, 1, 1)]]
+  [[intel::max_global_work_dim(3)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjGood7 {
+  [[cl::reqd_work_group_size(4, 1, 1)]]
+  [[intel::max_global_work_dim(3)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjGood8 {
+  [[intel::max_work_group_size(8, 1, 1)]]
+  [[intel::max_global_work_dim(3)]] void
+  operator()() const {}
+};
+
 #ifdef TRIGGER_ERROR
+// Checks correctness of mutual usage of different work_group_size attributes:
+// reqd_work_group_size, max_work_group_size and max_global_work_dim.
+// In case the value of 'max_global_work_dim' attribute equals to 0 we shall
+// ensure that if max_work_group_size and reqd_work_group_size attributes exist,
+// they hold equal values (1, 1, 1).
+
 struct TRIFuncObjBad {
   [[intel::max_global_work_dim(0)]]
   [[intel::max_work_group_size(8, 8, 8)]] // expected-error{{'max_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
   [[cl::reqd_work_group_size(4, 4, 4)]]   // expected-error{{'reqd_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
   void
   operator()() const {}
 };
+
+struct TRIFuncObjBad1 {
+  [[intel::max_work_group_size(8, 8, 8)]] // expected-error{{'max_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjBad2 {
+  [[cl::reqd_work_group_size(4, 4, 4)]]   // expected-error{{'reqd_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjBad3 {
+  [[intel::reqd_work_group_size(4, 4, 4)]]   // expected-error{{'reqd_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
```

Comment:
We silently accepted those test cases. With new design, the test cases generate expected diagnostic.